predicators/nsrt_learning/strips_learning/effect_search_learner.py

```diff
@@ -0,0 +1,424 @@
+        predicates: Set[Predicate],
+        segmented_trajs: List[List[Segment]],
+        effect_sets_to_pnads: Callable[[_EffectSets],
+                                       List[PartialNSRTAndDatastore]],
+        backchain: Callable[
+            [List[Segment], List[PartialNSRTAndDatastore], Set[GroundAtom]],
+            _Chain],
+    ) -> None:
+        self._trajectories = trajectories
+        self._train_tasks = train_tasks
+        self._predicates = predicates
+        self._segmented_trajs = segmented_trajs
+        self._effect_sets_to_pnads = effect_sets_to_pnads
+        self._backchain = backchain
+
+    @abc.abstractmethod
+    def __call__(self, effect_sets: _EffectSets) -> float:
+        """Compute the heuristic value for the given effect sets."""
+        raise NotImplementedError("Override me!")
+
+
+class _BackChainingHeuristic(_EffectSearchHeuristic):
+    """Counts the number of transitions that are not yet covered by some
+    operator in the backchaining sense."""
+
+    def __call__(self, effect_sets: _EffectSets) -> float:
+        pnads = self._effect_sets_to_pnads(effect_sets)
+        uncovered_transitions = 0
+        for ll_traj, seg_traj in zip(self._trajectories,
+                                     self._segmented_trajs):
+            if ll_traj.is_demo:
+                traj_goal = self._train_tasks[ll_traj.train_task_idx].goal
+                _, chain, _ = self._backchain(seg_traj, pnads, traj_goal)
+                assert len(chain) <= len(seg_traj)
+                uncovered_transitions += len(seg_traj) - len(chain)
+        return uncovered_transitions
+
+
+class EffectSearchSTRIPSLearner(BaseSTRIPSLearner):
+    """Base class for a effect search STRIPS learner."""
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "effect_search"
+
+    def _learn(self) -> List[PartialNSRTAndDatastore]:
+
+        # Set up hill-climbing search over effect sets.
+        # Create the search operators.
+        search_operators = self._create_search_operators()
+
+        # Create the heuristic.
+        heuristic = self._create_heuristic()
+
+        # Initialize the search.
+        initial_state = self._create_initial_effect_sets()
+
+        def get_successors(
+            effs: _EffectSets
+        ) -> Iterator[Tuple[Tuple[_EffectSearchOperator, int], _EffectSets,
+                            float]]:
+            for op in search_operators:
+                for i, child in enumerate(op.get_successors(effs)):
+                    yield (op, i), child, 1.0  # cost always 1
+
+        # Run hill-climbing search.
+        path, _, cost = utils.run_hill_climbing(initial_state=initial_state,
+                                                check_goal=lambda _: False,
+                                                get_successors=get_successors,
+                                                heuristic=heuristic)
+
+        # This effectively asserts harmlessness.
+        assert cost[-1] == 0
```

Comment:
i'm not sure if we'll keep this long-term because we might add some kind of regularization to the cost to encourage simpler operator sets. but so long as we are doing this, we might as well add `early_termination_heuristic_thresh=0` to `run_hill_climbing`, which will terminate the search as soon as we find an effect set that has 0 cost.